File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps):
    # 맵의 행과 열을 구한다.
    n, m = len(maps), len(maps[0])
    # 상하좌우
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    # BFS 탐색을 시작하기 위해 큐를 초기화
    queue = deque([(0, 0)])
    
    # queue가 빌 때까지 반복
    while queue:
        # x,y 나눠서 popleft() 해줘야
        x, y = queue.popleft()
        # 상하좌우 칸 확인하기
        for i in range(4):
            # 현재 위치에서 상하좌우로 이동 가능한 위치를 계산합니다. 
            nx, ny = x + dx[i], y + dy[i]
            # 맵을 벗어나면 무시하기
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            # 벽이면 무시하기
            if maps[nx][ny] == 0:
                continue
            # 처음 지나가는 길이면 거리계산하고 다시 상하좌우 확인하기
            if maps[nx][ny] == 1:
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx, ny))
    
    # 상대 팀 진영(제일 오른쪽 아래 칸)까지의 거리 반환
    return maps[n-1][m-1]

# solution 함수
def solution(maps):
    # bfs를 호출하고 결과를 바탕으로 정답을 반환
    answer = bfs(maps)
    #상대 팀 진영에 도착할 수 없을 때 -1(벽에 둘러쌓여 1이되어 도착하지 못함)
    if answer == 1 :
        return -1  
    else :
        return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.ArrayDeque;

public class Solution {
    // 두 팀으로 나누어서 진행하며 상대 팀 진영을 먼저 파괴하면 이기는 게임
    // 각 팀은 상대 팀 진영에 최대한 빨리 도착해야 게임을 유리하게

    // 너비 우선 탐색은 가중치가 없는 그래프에서 최단 경로를 보장합니다.
    // 현재 이 문제는 지나가는 칸의 최솟값을 구해야하기 때문에 너비 우선 탐색 사용

    private static final int[] rx = {0,0,-1,1};
    private static final int[] ry = {-1,1,0,0};

    private static class Node{
        int r,c;

        public Node(int r, int c) {
            this.r = r;
            this.c = c;
        }
    }

    public int solution(int[][] maps){
        // 맵의 크기를 저장하는 변수 선언
        int N = maps.length;
        int M = maps[0].length;

        // 최단 거리를 저장할 배열 생성
        int[][] dist = new int[N][M];

        //bfs 탐색을 위한 큐 생성
        ArrayDeque<Node> queue = new ArrayDeque<>();
        // 시작 정점에 대해서 큐에 추가, 최단 거리 저장
        queue.addLast(new Node(0,0));
        dist[0][0] = 1;

        // queue 가 빌 때까지 반복
        while (!queue.isEmpty()) {
            Node now = queue.pollFirst();

            // 현재 위치에서 이동할 수 있는 모든 방향
            for (int i = 0; i < 4; i++) {
                int nr = now.r + rx[i];
                int nc = now.c + ry[i];

                // 맵 밖으로 나가는 경우 예외처리
                if (nr < 0 || nc < 0 || nr >= N || nc >= M)
                    continue;

                // 벽으로 가는 경우 예외처리
                if (maps[nr][nc] == 0)
                    continue;

                // 이동한 위치가 처음 방문하는 경우, queue에 추가하고 거리 갱신
                if (dist[nr][nc] == 0) {
                    queue.addLast(new Node(nr, nc));
                    dist[nr][nc] = dist[now.r][now.c] + 1;
                }
            }
        }

        // 목적지까지 최단 거리 반환, 목적지에 도달하지 못한 경우에는 -1 반환
        if (dist[N - 1][M - 1] == 0) {
            return -1;
        } else {
            return dist[N - 1][M - 1];
        }
    }
}